data/transactions/logic/eval.go

```diff
@@ -2089,337 +2224,366 @@ func (cx *EvalContext) txnFieldToStack(txn *transactions.Transaction, fs txnFiel
 		sv.Uint = boolToUint(txn.AssetFrozen)
 	case ExtraProgramPages:
 		sv.Uint = uint64(txn.ExtraProgramPages)
+
+	case Logs:
+		if arrayFieldIdx >= uint64(len(stxn.EvalDelta.Logs)) {
+			err = fmt.Errorf("invalid Logs index %d", arrayFieldIdx)
+			return
+		}
+		sv.Bytes = nilToEmpty([]byte(stxn.EvalDelta.Logs[arrayFieldIdx]))
+	case NumLogs:
+		sv.Uint = uint64(len(stxn.EvalDelta.Logs))
+	case LastLog:
+		if logs := len(stxn.EvalDelta.Logs); logs > 0 {
+			sv.Bytes = nilToEmpty([]byte(stxn.EvalDelta.Logs[logs-1]))
+		} else {
+			sv.Bytes = nilToEmpty(nil)
+		}
+	case CreatedAssetID:
+		sv.Uint = uint64(stxn.ApplyData.ConfigAsset)
+	case CreatedApplicationID:
+		sv.Uint = uint64(stxn.ApplyData.ApplicationID)
+
 	default:
-		err = fmt.Errorf("invalid txn field %d", fs.field)
+		err = fmt.Errorf("invalid txn field %s", fs.field)
 		return
 	}
 
-	txnFieldType := TxnFieldTypes[fs.field]
-	if !typecheck(txnFieldType, sv.argType()) {
-		err = fmt.Errorf("%s expected field type is %s but got %s", fs.field.String(), txnFieldType.String(), sv.argType().String())
+	if !typecheck(fs.ftype, sv.argType()) {
+		err = fmt.Errorf("%s expected field type is %s but got %s", fs.field, fs.ftype, sv.argType())
 	}
 	return
 }
 
-func opTxn(cx *EvalContext) {
-	field := TxnField(cx.program[cx.pc+1])
+func (cx *EvalContext) fetchField(field TxnField, expectArray bool) (*txnFieldSpec, error) {
 	fs, ok := txnFieldSpecByField[field]
 	if !ok || fs.version > cx.version {
-		cx.err = fmt.Errorf("invalid txn field %d", field)
-		return
+		return nil, fmt.Errorf("invalid txn field %d", field)
 	}
-	_, ok = txnaFieldSpecByField[field]
-	if ok {
-		cx.err = fmt.Errorf("invalid txn field %d", field)
-		return
+	if expectArray != fs.array {
+		if expectArray {
+			return nil, fmt.Errorf("unsupported array field %d", field)
+		}
+		return nil, fmt.Errorf("invalid txn field %d", field)
+	}
+	return &fs, nil
+}
+
+type txnSource int
+
+const (
+	srcGroup      txnSource = iota
+	srcInner                = iota
+	srcInnerGroup           = iota
+)
+
+// opTxnImpl implements all of the txn variants.  Each form of txn opcode should
+// be able to get its work done with one call here, after collecting the args in
+// the most straightforward way possible. They ought to do no error checking, so
+// that it is all collected here.
+func (cx *EvalContext) opTxnImpl(gi uint64, src txnSource, field TxnField, ai uint64, expectArray bool) (sv stackValue, err error) {
```

Comment:
variable type inconsistency : the `gi` variable would probably be of an `int` type here.
while I think that would be better to have it as `uint64` everywhere, it looks like this is the only exception - so, let's make this align with the rest of the codebase ?

Reply:
This was intentional, as described in the comment above. opTxnImpl takes "wide" types so that it can be called safely from the various txn opcodes, some of which have bytes for gi, some have uint64 (since they got the index off the stack).  This is the one place ion the code where we confirm that these indexes fit into ints (actually, we confirm much more - that they are small enough for the txn group). And so from here on, they can be coerced into `int`.

The same is true of `ai`